dropshot/src/server.rs

```diff
@@ -173,6 +152,85 @@ impl HttpServer {
     }
 }
 
+/**
+ * A running Dropshot HTTP server.
+ */
+pub struct HttpServer {
+    app_state: Arc<DropshotState>,
+    local_addr: SocketAddr,
+    join_handle: Option<tokio::task::JoinHandle<Result<(), hyper::Error>>>,
+    close_channel: Option<tokio::sync::oneshot::Sender<()>>,
+}
+
+impl HttpServer {
+    pub fn local_addr(&self) -> SocketAddr {
+        self.local_addr
+    }
+
+    pub fn app_private(&self) -> Arc<dyn Any + Send + Sync + 'static> {
+        Arc::clone(&self.app_state.private)
+    }
+
+    /**
+     * Signals the currently running server to stop (if it hasn't
+     * stopped already) and waits for it to exit.
+     */
+    pub async fn close(mut self) -> Result<(), String> {
+        self.close_channel
+            .take()
+            .expect("cannot close twice")
+            .send(())
+            .expect("failed to send close signal");
+        if let Some(handle) = self.join_handle.take() {
```

Comment:
When exactly would we get here and find that this was not `Some`?

Reply:
If the implementation of `Future` returns `Ready`.

Basically, in the implementation of `Future`, we poll the handle. If it completes, we drop it, to avoid re-polling a completed future (which is part of the Future API requirements: "[Once a future has finished, clients should not poll it again.](https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll)").

In other word: `self.join_handle` should be `Some` if we haven't seen the future complete, and should be `None` if it *has* completed.

Reply:
If we drop it in the Future, how can we get to this point in `close` and find it `None`?

Reply:
> If we drop it in the Future

Yeah, so "drop it in the Future" means (referring to my implementation of `HttpServer::poll`):

- We called `server.join_handle.take()` to access the handle,
- We polled the handle and it was not pending
- We `return poll` without "putting the handle back in `self.join_handle`", which we would have done if `poll.is_pending()` was true.

> how can we get to this point in `close` and find it `None`?

`self.join_handle` is an `Option`. In the implementation of `Future`, we **always** call: `server.join_handle.take()`, and we **only** put the handle back if `poll.is_pending()`.

So *if* we dropped it in the future (by completing the join_handle), close will *always* find it `None`, because the implementation of `poll` called `take()` without `replace()`.

Does that make sense?

Reply:
Got it -- we drop the handle, not the server.  So the sequence of events that gets us here is just that someone has waited for the `Future` and then called `close()`.  Thanks for clarifying.

Reply:
Exactly!